### PR TITLE
Connorwon 237 update backend security

### PR DESCRIPTION
**Pull request by ConnorWon**
## Description

Refactoring implementation of backend api access

### Changes

Deleted API key authorization
Refactored our implementation of access token
- converted it to a JWT token
- removed it from User Model
- send it to the front end via cookie
- added middleware that runs on all non-GET endpoints to verify user is authorized

Refactored implementation of refresh token,  and created separate db collection for it
- removed form User model & created separate db collection for it
- sent to the frontend via cookie

**NOTE: Add JWT_SECRET_KEY=thisisthesecretkey to your .env file**

## Checklist

- [x] Ran `npm run lint:fix` to format code
- [x] Assigned PR to all authors
- [x] Requested at least one reviewer

### CODE REVIEW

**Comment by ConnorWon** [comment]
[Update Security for Backend](https://trello.com/c/RcrTUCq1/237-update-security-for-backend)

**Comment by ElisaHu1** [comment]
this is such a major change to backend, i wonder what you have tested?

**Comment by ConnorWon** [comment]
> this is such a major change to backend, i wonder what you have tested?

Token-focused Testing
- Logging in → create access token & refresh token
- Logout → token cookies are deleted and refresh token is deleted from db
- Closing browser → going back to website → user is automatically logged back in (via refreshToken controller endpoint)
- Access token expires → make a call that requires authentication (ex. delete a resource) → access token is reacquired
- Access token expires (no refresh token) → make a call that requires authentication (ex. delete a resource) → authenticate token middleware fails

Other Tests
- Creating a resource
- Deleting a resource
- Viewing resources while logged in
- Viewing resource while not logged in
- Viewing people page
- Viewing project team page
- Creating a new member via the Project team page
- Logging in and out on pages that have components that are visible/not visible based on login status

I also ran all the unit tests

**Comment by ConnorWon** [comment]
Also, forgot to mention but I added two new libraries to the backend, so you'll have to update your dependencies for it to work properly

**Comment by ElisaHu1** [comment]
image
backend not working

**Comment by ConnorWon** [comment]
> image backend not working

This should be fixed with a commit I pushed in the frontend

File: controllers/authController.js [hunk 1]
```diff
@@ -202,38 +153,28 @@ export const loginUser = async (req, res) => {
...
-			});
-		} else {
-			res.status(400).json({
-				message: 'Invalid refresh token.',
-			});
+		if (!refresh_token) {
+			return res.sendStatus(400);
 		}
+
+		const access_token = sign({id: refresh_token.user._id, username: refresh_token.user.username, permissions: refresh_token.user.permissions}, process.env.JWT_SECRET_KEY, {expiresIn: '1h'});
```
Comment (ElisaHu1) [inline controllers/authController.js:168]: just curious, why 1h? also this is a dirty number, can you make a constant instead?

**Review by ryanyae — APPROVED** [review]
Looks good to me

File: controllers/authController.js (hunk 1 above)
Comment (ConnorWon) [reply controllers/authController.js:168]: Pretty much the shorter the "expiresIn" length the greater security, but obviously the shorter the length the more times we'll have to call refreshToken. So taking that into account as well as the fact that the information we are protecting via authentication is not extremely important to protect, I came to the length of 1 hour. However, we can always change this if we want to as its a simple change.

**Review by ElisaHu1 — COMMENTED** [review]
looks good, can you clean up the legacy code? (instead of comment out) also add more comments for your design when necessary across the file

**Review by ElisaHu1 — APPROVED** [review]

**Pull request merged**